packages/loopring_v3/test/testExchangeUtil.ts

```diff
@@ -1948,6 +2037,169 @@ export class ExchangeTestUtil {
+  ) {
+    let pendingTransferres = this.pendingInternalTransferRequests[exchangeID];
+
+    if (pendingTransferres.length === 0) {
+      return;
+    }
+
+    const blockType = BlockType.INTERNAL_TRANSFER;
+
+    let numTransferDone = 0;
+    while (numTransferDone < pendingTransferres.length) {
+      const transferres: InternalTransferRequest[] = [];
+      let numRequestsInBlock = 0;
+      // Get all withdrawals for the block
+      const blockSizes = this.interfalTransferBlockSizes;
+      const blockSize = this.getBestBlockSize(
+        pendingTransferres.length - numTransferDone,
+        blockSizes
+      );
+      for (let b = numTransferDone; b < numTransferDone + blockSize; b++) {
+        if (b < pendingTransferres.length) {
+          // pendingTransferres[b].slotIdx = withdrawals.length;
+          transferres.push(pendingTransferres[b]);
+          numRequestsInBlock++;
+        } else {
+          const dummyInternalTransferresRequest: InternalTransferRequest = {
+            accountFromID: this.dummyAccountId,
+            accountToID: this.dummyAccountId,
+            transTokenID: 0,
+            amount: new BN(0),
+            feeTokenID: 0,
+            fee: new BN(0),
+            label: 0
+          };
+          transferres.push(dummyInternalTransferresRequest);
+        }
+      }
+      assert(transferres.length === blockSize);
+      numTransferDone += blockSize;
+
+      // Hash the labels
+      const labels: number[] = [];
+      for (const trans of transferres) {
+        labels.push(trans.label);
+      }
+      const labelHash = this.hashLabels(labels);
+
+      // Sign the offchain withdrawals
+      for (const trans of transferres) {
+        this.signInternalTransfer(trans);
+      }
+
+      // Block info
+      const operator = await this.getActiveOperator(exchangeID);
+      const internalTransferBlock: InternalTransferBlock = {
+        transferres,
+        onchainDataAvailability: this.onchainDataAvailability,
+        operatorAccountID: operator
+      };
+
+      // Store state before
+      const currentBlockIdx = (await this.exchange.getBlockHeight()).toNumber();
+      const stateBefore = await this.loadExchangeState(
+        exchangeID,
+        currentBlockIdx
+      );
+
+      const jWithdrawalsInfo = JSON.stringify(
+        internalTransferBlock,
+        replacer,
+        4
+      );
+      const [blockIdx, blockFilename] = await this.createBlock(
+        exchangeID,
+        blockType,
+        jWithdrawalsInfo
+      );
+
+      // Store state after
+      const stateAfter = await this.loadExchangeState(
+        exchangeID,
+        currentBlockIdx + 1
+      );
+
+      const block = JSON.parse(fs.readFileSync(blockFilename, "ascii"));
+      const bs = new Bitstream();
+      bs.addNumber(block.exchangeID, 4);
+      bs.addBN(new BN(block.merkleRootBefore, 10), 32);
+      bs.addBN(new BN(block.merkleRootAfter, 10), 32);
+      for (const trans of block.internalTransferres) {
```

Comment:
Not sure why this is here as well as below. The only non-DA data is `exchangeID`, `merkleRootBefore`, `merkleRootAfter` and `labelhash` (though this value is also ignored on-chain, but we publish it always anyway because the cost is insignificant and constant for a block). Nothing needs to be stored on-chain for an offchain transfer (so this is different compared to a withdrawal because a user needs to be able to withdraw from the contract using on-chain data).

Reply:
Hi @Brechtpd, Thanks for the review! In fact, I have a question here: how to choose the DA, any criteria? I selected these members just like off-chain withdraw to quick enable/test my transfer circuit implementation. The value I picked somehow like a combination of ApprovedWithdrawalData and DataAvailabilityData in OffchainWithdrawGadget.
And here I have another question, what's the purpose of ApprovedWithdrawalData?
Thanks!

Reply:
> I have a question here: how to choose the DA, any criteria? I selected these members just like off-chain withdraw to quick enable/test my transfer circuit implementation. The value I picked somehow like a combination of ApprovedWithdrawalData and DataAvailabilityData in OffchainWithdrawGadget.

It's the minimal amount of data so that you can create new Merkle tree from the older Merkle tree + the DA data. So I think the data you've selected is correct.

> And here I have another question, what's the purpose of ApprovedWithdrawalData?

This data will be stored on-chain. It's important that this is stored on-chain because:
- A user can request a withdrawal on the contract, so this data needs to be available in the contract.
- This is needed for the 2 step `commitBlock`/`verifyBlocks`. This data will be stored on-chain in `commitBlock`, but the data can only be used by a user for a withdrawal once the block is finalized (otherwise the state is not guaranteed correct. An operator could commit an invalid block with too large a withdrawal amount, but the operator will not be able to create a proof for a block like this).